src/Nri/Ui/Carousel/V2.elm

```diff
@@ -0,0 +1,307 @@
+
+-}
+
+import Accessibility.Styled.Aria as Aria
+import Accessibility.Styled.Role as Role
+import Css exposing (..)
+import Html.Styled as Html exposing (..)
+import Html.Styled.Attributes as Attrs exposing (css, id)
+import List.Extra
+import Maybe.Extra
+import Nri.Ui.ClickableSvg.V2 as ClickableSvg
+import Nri.Ui.Svg.V1 exposing (Svg)
+import TabsInternal.V2 as TabsInternal
+
+
+{-| Type which represents the type of aria label which will be used
+`LabelledByIdOfVisibleLabel` will point to an existing element id on the DOM
+`LabelledByAccessibleLabelOnly` will be a label of the element
+-}
+type LabelledBy
+    = LabelledByIdOfVisibleLabel String
+    | LabelledByAccessibleLabelOnly String
+
+
+{-| `Role`, which can be either [Group](https://w3c.github.io/aria/#group) or [Region](https://w3c.github.io/aria/#region)
+-}
+type Role
+    = Group
+    | Region
+
+
+{-| Builds a carousel with previous and next controls
+Returns:
+`slides` the container with the carousel contents
+`viewPreviousButton` previous button
+`viewNextButton` next button
+`containerAttributes` attributes that should be used on the parent div of both the button and slides elements
+-}
+viewWithPreviousAndNextControls :
+    { selected : id
+    , panels :
+        List
+            { id : id
+            , slideHtml : Html msg
+            , labelledBy : LabelledBy
+            , idString : String
+            }
+    , viewPreviousButton : { attributes : List (ClickableSvg.Attribute msg), icon : Svg, name : String }
+    , viewNextButton : { attributes : List (ClickableSvg.Attribute msg), icon : Svg, name : String }
+    , labelledBy : LabelledBy
+    , role : Role
+    , focusAndSelect : { select : id, focus : Maybe String } -> msg
+    }
+    ->
+        { viewPreviousButton : Html msg
+        , viewNextButton : Html msg
+        , slides : Html msg
+        , containerAttributes : List (Attribute msg)
+        }
+viewWithPreviousAndNextControls config =
+    let
+        currentPanelIndex =
+            List.Extra.findIndex (\p -> p.id == config.selected) config.panels
+
+        previousPanel =
+            currentPanelIndex
+                |> Maybe.andThen
+                    (\index ->
+                        List.Extra.getAt
+                            (if index - 1 >= 0 then
+                                index - 1
+
+                             else
+                                List.length config.panels - 1
+                            )
+                            config.panels
+                    )
+
+        nextPanel =
+            currentPanelIndex
+                |> Maybe.andThen
+                    (\index ->
+                        List.Extra.getAt
+                            (if index + 1 < List.length config.panels then
+                                index + 1
+
+                             else
+                                0
+                            )
+                            config.panels
+                    )
```

Comment:
Could you add tests, please? I'm thinking that this logic in particular would be nice to have covered, by way of some program-tests with varying numbers of slides.